### PR TITLE
Trigger build workflow via PR comment instead of PR state; add a workflow to build an updated SDK container, 

### PR DESCRIPTION
**Pull request by t-lo**
This change adds a github actions workflow to trigger SDK and OS image builds from PR comments. It verifies whether the commenter is a member of the [Flatcar Maintainers](https://github.com/orgs/flatcar/teams/flatcar-maintainers) team before commencing the build.

The PR comment commands are:
- `/build-sdk` - builds SDK and OS image, runs tests
- `/build-image` - builds OS image and runs tests

The previous semantics of triggering builds from PR changes ("ready to review", "synchronised") is removed. **This implies builds now need to be explicitly started using one of the above commands.**

Further, a new workflow is added to create a new SDK container based on an existing SDK container. The new container image is published to bincache. This can be used for CI testing intrusive changes that also affect the SDK without bootstrapping a whole new SDK.

A garbage collector is added to `ci-automation`  for cleaning up SDK containers on bincache.

# Testing done

- Maintainer team membership with hard-coded "commenter" names
  - pass: user `t-lo` - https://github.com/flatcar/scripts/actions/runs/4916044777
  - fail: user `mallory` - https://github.com/flatcar/scripts/actions/runs/4915901180

Further testing is challenging before we merge since the corresponding events (PR comment, PR-has-changed) are only emitted to workflows in the "main" branch.

### CODE REVIEW

**Comment by t-lo** [comment]
Rebased on latest main.

File: .github/workflows/pr-comment-build-dispatcher.yaml
Comment (t-lo) [reply .github/workflows/pr-comment-build-dispatcher.yaml:65]: Good catch! Will fix.

**Comment by t-lo** [comment]
Note: Test results can be ignored because some parts of the workflows will only work after merge to main. This is caused by Github only emitting certain action events to workflows in the main repo.

**Comment by pothos** [comment]
I don't understand why we would need to publish a container image. The SDK container could be built as part of `ci.yaml` and doesn't need to be uploaded and downloaded.
Since it doesn't bootstrap a new SDK but only updates packages, this is fast - and I wonder if we even need to build this new container image, I think we should rather call `./update_chroot --toolchain_boards="amd64-usr arm64-usr"` and `./setup_board --board="arm64-usr" --regen_configs && ./setup_board --board="amd64-usr" --regen_configs` as optional preparation step in `ci.yaml`.

**Comment by pothos** [comment]
> We actually perform the above steps [every time the SDK container is entered](https://github.com/flatcar/scripts/blob/main/sdk_lib/sdk_entry.sh#L40-L43). So in that case, one would simply go with `/build-image`.

Maybe `update_chroot` does the difference or that some directory is not there in the container image creation? We could also do these steps in `ci.yaml` (optional if they slow the build down). I don't understand the use case of building a new SDK this way; if it would be fully bootstrapped, then yes.

**Comment by t-lo** [comment]
> > We actually perform the above steps [every time the SDK container is entered](https://github.com/flatcar/scripts/blob/main/sdk_lib/sdk_entry.sh#L40-L43). So in that case, one would simply go with `/build-image`.
> 
> Maybe `update_chroot` does the difference or that some directory is not there in the container image creation? We could also do these steps in `ci.yaml` (optional if they slow the build down). I don't understand the use case of building a new SDK this way; if it would be fully bootstrapped, then yes.

The only reason why we would do this step is to explicitly produce a new SDK container image for later re-use (manual testing, bugfixing, reproduction of issues) with the changes proposed by the PR. As stated, we're doing the same *transiently* in the package build. The only benefit of calling `/build-sdk` instead of `/build-image` is to generate a separate artifact. Which might be useful for debugging downstream build issues (think auto-cut toolchain update PRs etc.).

Re: SDK bootstrapping, that's a good point - we can consider adding this at a later point.

**Comment by pothos** [comment]
For me the benefit and use case is a bit too narrow compared to running this command on the nightly SDK when needed. As high level action, the full bootstrap would be a better fit. This doesn't have to be done now but later I would like to see that `/build-sdk` does a full bootstrap because it's the more common thing to do and covers this special case, too, with a bit more waiting. As optimization this reusing of an existing SDK makes sense and could be offered through a special parameter or new command, similar to how we could offer a two-phase SDK build.

**Comment by t-lo** [comment]
Makes sense - I'll change the command to `/update-sdk` for the time being so we can add `/build-sdk` later.

**Comment by t-lo** [comment]
Done. Coincidentally the actual workflow file's name already was `update-sdk.yaml`.

File: .github/workflows/pr-comment-build-dispatcher.yaml
Comment (pothos) [inline .github/workflows/pr-comment-build-dispatcher.yaml:31]:
```suggestion
          set -euo pipefail
          curl --fail --show-error -L --silent \
```
When the API fails, we can exit with an error instead of having an empty maintainers file

File: .github/workflows/update-sdk.yaml
Comment (pothos) [inline .github/workflows/update-sdk.yaml:130]: There's some risk attached as GitHub Actions could be misconfigured to allow extraction when it runs without manual approval required. Another option would be to save this as regular CI artifact. What do you think?

Comment (pothos) [reply .github/workflows/update-sdk.yaml:130]: (Plus we don't need to think about extra garbage collection)

Comment (t-lo) [reply .github/workflows/update-sdk.yaml:130]: I understand the concern and I realise it's a trade-off.
My concerns are that using github artifacts will make the overall process a lot slower - a lot of time is already spent on copying back and forth image build artifacts. Bincache access on the other hand is  significantly faster - often by a factor of 10 or higher.
Also, it's not unlikely that - even though officially Artifacts storage is unlimited for public repos - we'd eventually create friction with Github since we'd potentially produce hundreds of gigabytes of artifacts per week.

Comment (pothos) [reply .github/workflows/update-sdk.yaml:130]: Ok, maybe I don't know all plans, I thought that building an SDK is a rare even and while bigger than the regular images, it's still not that much of a difference.

Comment (t-lo) [reply .github/workflows/update-sdk.yaml:130]: At the end of the day this is just one building block for fully automating merging of auto-generated PRs.
I think quite a few of the PRs that we're generating currently - kernel, rust, bulk Gentoo packages refreshes - would mandate an SDK update if we really want to test these fully automatedly.

But you're right with your concern, and one way around it would be to run the SDK container update as an optional step in the main `ci.yaml`. I'll think through it.

Comment (pothos) [reply .github/workflows/update-sdk.yaml:130]: Adding support for the SDK tar ball as GitHub Action artifact would be interesting security-wise but I think we can merge it as is and maybe try this out in a follow-up (I don't know how easy that would be).

**Review by pothos — APPROVED** [review]

**Review by dongsupark — APPROVED** [review]

File: ci-automation/garbage_collect_github_ci_sdk.sh
Comment (krnowak) [reply ci-automation/garbage_collect_github_ci_sdk.sh:58]: Although safer would probably be define a new variable (`keep_tail`) as we seem to be using the `keep` variable also for `awk` invocation.

File: ci-automation/garbage_collect_github_ci_sdk.sh
Comment (krnowak) [inline ci-automation/garbage_collect_github_ci_sdk.sh:72]: If you decide to have a variable `keep_tail` above, thus keeping `keep` unmodified, then I think that the ` - 1` could be removed.

File: ci-automation/garbage_collect_github_ci_sdk.sh
Comment (krnowak) [inline ci-automation/garbage_collect_github_ci_sdk.sh:60]: I'd say it usually better to keep list of values as arrays, especially when you want to iterate it with a loop later. Unfortunately bash has a quite obtuse way of creating arrays from multiline strings:

```suggestion
    local purge_versions
    mapfile -t purge_versions < <(tail -n+"${keep}" <<<"${versions_sorted}")
```

Comment (krnowak) [reply ci-automation/garbage_collect_github_ci_sdk.sh:60]: Or, if you'll go with `tail_keep` variable:

```suggestion
    local purge_versions
    mapfile -t purge_versions < <(tail -n+"${tail_keep}" <<<"${versions_sorted}")
```

File: ci-automation/garbage_collect_github_ci_sdk.sh
Comment (krnowak) [reply ci-automation/garbage_collect_github_ci_sdk.sh:72]: If `purge_versions` is an array then this would need to be something like:
```suggestion
    printf '%s\n' "${purge_versions[@]}" | awk -v keep="${keep}" '{if ($0 == "") next; printf "%5d %s\n", NR + keep - 1, $0}'
```

File: ci-automation/garbage_collect_github_ci_sdk.sh
Comment (krnowak) [inline ci-automation/garbage_collect_github_ci_sdk.sh:77]: If `purge_versions` is an array then this would need to be:

```suggestion
    for version in "${purge_versions[@]}"; do
```

**Review by krnowak — COMMENTED** [review]
Some nitpicks.

**Comment by t-lo** [comment]
I've updated `garbage_collect_github_ci_sdk.sh` as suggested by @krnowak in https://github.com/flatcar/scripts/pull/800#discussion_r1196572965; we're now using a separate variable `tail_keep` for the `tail` invocation.

**Review by krnowak — APPROVED** [review]

**Pull request merged**
